### ISSUE TITLE
Provide an option to disable fallback

### ISSUE THREAD

**Issue by lessload** [AdguardTeam/AdguardForAndroid#3447]
### Issue Details

- **AdGuard version:**
  - v3.4.120
- **Filtering mode:**
  - Local VPN, DNSCrypt (Adguard)
- **Device:**
  - Mi CC9 Pro (Xiaomi)
- **Operating system and version:**
  - Android 10
- **Root access:**
  - No

### Expected Behavior

App should always filtering traffic or have something like Kill Switch. and stop only when disable protection. 

### Actual Behavior

After update to v3.4. I found dns traffic leak to my pi-hole randomly but too many dns leak. Adguard show as protection enable but traffic still leak need to re-enable to fix it. i always active Adguard, and enable Always-on VPN function in android setting. The traffic that leak not show in Filtering log. Please check.

### Additional Information

[already set battery - No restrictions / enabled - auto start] but when update blocklist it couldn't leak the traffic. it shoult have something like kill switch to protect the traffic.
Adguard not work with multi-user android? (dual-app on MIUI)

**Comment by lessload** [comment]
As several test it seem like the problem was from DNSCrypt only. not sure why. Not found such a problem on DoH.

**Comment by lessload** [comment]
Long time test nothing can fix. not sure the problem was from. now go back to 3.3.3. Not found such a problem.

**Comment by ameshkov** [comment]
@lessload is there any change when you select DNS-over-HTTPS traffic?

**Comment by lessload** [comment]
> 
> 
> @lessload is there any change when you select DNS-over-HTTPS traffic?

it's still leak.

**Comment by the4anoni** [comment]
I've got dns leaks too. 
Galaxy S10 OneUI 2.1

**Comment by the4anoni** [comment]
> @lessload is there any change when you select DNS-over-HTTPS traffic?

I used Opendns DoH in adguard

**Comment by Chinaski1** [comment]
To troubleshoot the issue, I would ask you to collect the following information:

- AdGuard > Settings > Advanced > Logging Level > Record All.
- Reproduce your issue.
- Note the time the error occurred and let us know.
- Don’t forget to return to the default logging level.
- Logs can be sent on devteam@adguard.com

**Comment by akmLaVx** [comment]
Logs received from 2520013.

**Comment by the4anoni** [comment]
> I've got dns leaks too.
> Galaxy S10 OneUI 2.1
> 
> Leak was from mobilewips app com.samsung.android.server.wifi.mobilewips.client

My mum Galaxy S9 with latest AdGuard 3.5 nightly 3 also has dns leaks. 
Sample leaked domains:
lh4.googleusercontent.com
connectivitycheck.gstatic.com
dls-udc.dqa.samsung.com

**Comment by cute-byte** [comment]
> I've got dns leaks too.
> Galaxy S10 OneUI 2.1
> 
> Leak was from mobilewips app com.samsung.android.server.wifi.mobilewips.client

Подтверждаю

**Comment by Ayan3199** [comment]
Got dns leak from adguard 3.2.120, android 10 One UI 2.0

Steps to reproduce-
1. Turn off internet completely for android device
2. Restart device
3. After reboot start adguard vpn, grant root access. 
4. After vpn services have started, turn on the internet now.
5. Open chrome now.
6. Observe vpn logs. I found multiple DNS resolution request through system default dns.

Temporary fix.
1. Turn off internet. 
2. Stop adguard vpn and delete vpn profile for adguard
3. Switch back to adguard app and toggle the switch.
4. System OS requests vpn creation permission. Wait.
5. Turn on internet now and then grant vpn permission. 
6. The dns leak can't be seen anymore

A note:-Chrome bydefault uses Google DNS to resolve queries. We need to disable this feature before adguard can actually block chrome dns requests. Go to Chrome://flag. Search for DNS.  Disable what u see

**Comment by ameshkov** [comment]
Quick update on this.

1. AdGuard may fall back to the system DNS if the upstream DNS does not answer in 2 seconds which may happen.
2. This is intended and there are multiple reasons for making this the default behavior. Otherwise, AdGuard would break authentication in some hotspot Wi-Fi networks. Also, we don't want to break internet access when the upstream is not available for some reason.

You can change the fallback DNS server addresses in the low-level settings.

@artemiv4nov what we need to do in this task is providing an option to disable fallback via low-level settings.
For instance, we could change the default behavior when `pref.dns.fallback` is set to "none" -- explain this in the description.

How to test fallback:
1. Set upstream DNS to some non-existing DNS server
2. Check the filtering log -- what "Upstream" server is shown for DNS queries there
3. Disable fallback
4. Ensure that the internet does not work anymore

**Comment by Ayan3199** [comment]
Got dns leak from adguard 3.2.120, android 10 One UI 2.0

Steps to reproduce-

Turn off internet completely for android device
Restart device
After reboot start adguard vpn, grant root access.
After vpn services have started, turn on the internet now.
Open chrome now.
Observe vpn logs. I found multiple DNS resolution request through system default dns.
Temporary fix.

Turn off internet.
Stop adguard vpn and delete vpn profile for adguard
Switch back to adguard app and toggle the switch.
System OS requests vpn creation permission. Wait.
Turn on internet now and then grant vpn permission.
The dns leak can't be seen anymore
A note:-Chrome bydefault uses Google DNS to resolve queries. We need to disable this feature before adguard can actually block chrome dns requests. Go to Chrome://flag. Search for DNS. Disable what u see

**Comment by Ayan3199** [comment]
While I carried out the test 1.1.1.1 was perfectly reachable from my pc @60ms

**Comment by ameshkov** [comment]
> Observe vpn logs. I found multiple DNS resolution request through system default dns.

Could you please show a few records?

**Comment by techIndia-hacker** [comment]
Yeah I have sent 4 screen shots at support@adguard

**Comment by techIndia-hacker** [comment]
Another update. Chrome 84.0.4147.89 uses inbuilt DNS resolver activity and does not send resolution requests anymore.

Steps to reproduce my observations-
1) clear chrome data. Let's start clean🙂 to avoid cached dns
2) open adguard filtering log and now tap the chrome icon
3) initially few domain resolutions get logged like update.googleapis.com etc as u go though the initial setting options ( or u may not because of DNS request leak though system dns as discussed in previous comments)
4) wait a minute for initialization to complete and then type in any website like Sony.com or ti.com. just see that u don't have any app installed that also visits same website otherwise it will be confusing.
5) i didn't observe any domain resolution requests

**Comment by techIndia-hacker** [comment]
Ok breakthrough in experimentation 😅😅😇

Disabling "pref.vpn.disable.reconfigure"  in low level settings solves the DNS leak issue 100% according to my observations atleast on my phone (samsung android 10 one ui 2.0)

**Comment by lessload** [comment]
> 
> 
> Ok breakthrough in experimentation 😅😅😇
> 
> Disabling "pref.vpn.disable.reconfigure" in low level settings solves the DNS leak issue 100% according to my observations atleast on my phone (samsung android 10 one ui 2.0) for most apps like chrome etc
> 
> But some apps uses tcp://8.8.8.8:53 for dns. They r still leaking. These things are hardcoded into specific apps I guess.

I can accept some traffic leak from hardcoded. But can not accept when it leak while Adguard update itself (a lot of traffic leak show on Pi-Hole) (sorry if you already fix it, i still use old version)

For me, Block Google DNS make all (maybe) traffic go through Adguard app.

**Comment by techIndia-hacker** [comment]
> I can accept some traffic leak from hardcoded. But can not accept when it leak while Adguard update itself (a lot of traffic leak show on Pi-Hole) (sorry if you already fix it, i still use old version)

In the previous versions of adguard  vpn or proxy reconfiguration was not needed to prevent dns leak but in latest version it's the only fix I found [I am not in anyway linked to adguard development team].

But u r taking about leak during update which is quite normal as the vpn or proxy is killed and restarted shortly after update completion. Though this issue should be fixed, it was there from the very beginning. 


> For me, Block Google DNS make all (maybe) traffic go through Adguard app.
Yeah i tried that but websites load very slowly if done, according to my observations

**Comment by artemiv4nov** [comment]
Improved.
We added a special logic to the `pref.fallback.dns` property.
An empty string means using the system default DNS as a fallback. A string `none` means no fallback at all.

**Comment by techIndia-hacker** [comment]
Browsers are not using configured dns even in the latest nightly v7 if
fallback is set to none and adguard vpn is started without network. I could
not find any DNS resolution requests from chrome or Samsung browser in
adguard log. But this issue is resolved immediately if adguard vpn is
reconfigured with internet connected.

**Comment by techIndia-hacker** [comment]
See attachment and 49.45.0.1:53 is the cached DNS server of my LTE service
provider

**Comment by ameshkov** [comment]
@techIndia-hacker are you using auto+proxy filtering mode?

> See attachment and 49.45.0.1:53 is the cached DNS server of my LTE service
provider

I see no attachments

**Comment by techIndia-hacker** [comment]
No I am using regular vpn mode though i have root access.

3 png screenshots are attached to the previous mail. Was it not received
correctly?

**Comment by ameshkov** [comment]
> 3 png screenshots are attached to the previous mail. Was it not received correctly?

Nope, you need to attach them manually on Github

**Comment by techIndia-hacker** [comment]
The image may contain traceable data. So can i send u through mail?

**Comment by ameshkov** [comment]
Sure, send them to `devteam@adguard.com`

We may also need to get additional logs.

Here's what we need you to do:

1. Open AdGuard -> Settings -> Advanced -> Set "Logging level" to "Record everything". It's desirable to close all background apps that don't concern the problem beforehand.
2. Reproduce the issue and set the logging level back to "Default".
3. Remember the exact time when the issue was reproduced, we will need it to find the corresponding records in the log file.
4. Open AdGuard -> Settings -> Advanced -> Export Logs
5. Additionally, capture the system bug report after reproducing the issue. [Here's the instruction](https://developer.android.com/studio/debug/bug-report).
6. Send both files to `devteam@adguard.com`. Mention the Github issue number and the exact time when the issue was reproduced.

**Comment by techIndia-hacker** [comment]
Sent please check

**Comment by techIndia-hacker** [comment]
Dns leak still happening to me. Sent logs. How this got closed?

**Comment by artemiv4nov** [comment]
@techIndia-hacker Hi! please re-check on the latest nightly build.

**Comment by techIndia-hacker** [comment]
I have checked in latest nightly too. Only solution for me is to enable vpn
reconfiguration on network change

**Comment by artemiv4nov** [comment]
> I have checked in latest nightly too. Only solution for me is to enable vpn reconfiguration on network change
> […](#)

Did you try to disable fallback at all? I described this case [a few comments above](https://github.com/AdguardTeam/AdguardForAndroid/issues/3447#issuecomment-670921389)

**Comment by techIndia-hacker** [comment]
Yeah it's set to none

**Comment by artemiv4nov** [comment]
@techIndia-hacker Can you send the debug logs again, please? Can't find them

**Comment by techIndia-hacker** [comment]
Search in the devteam mail box for my name

**Comment by ameshkov** [comment]
@techIndia-hacker the latest logs are from the week ago, please collect the logs with the new nightly and send them to devteam.

**Comment by techIndia-hacker** [comment]
Ok will send u when i get time

**Comment by techIndia-hacker** [comment]
Sent logs

**Comment by artemiv4nov** [comment]
@techIndia-hacker Got it, thanks. We will investigate your logs tomorrow.

**Comment by admitrevskiy** [comment]
@techIndia-hacker Hi! 
This case looks like built-in Android Personal-DNS behavior. Please, check if it's enabled in system settings:

Settings -> Network -> Advanced -> Personal DNS. 
For new Android versions it's enabled by default and set to `Auto` mode. It should be disabled, our DNS will conflict with the built-in otherwise

**Comment by admitrevskiy** [comment]
BTW about plain-DNS leaks. if you want you can use DoT server with ip e.g. `tls://1.1.1.1` (if the server allows it) as a bootstrap for DoT and DoH servers: you can configure it in Advanced Settings -> `pref.dns.bootstrap`

Or you can use true DNSCrypt servers

But this refers to a single one bootstrap query to resolve DoH/DoT server address

**Comment by techIndia-hacker** [comment]
Private DNS is turned off even when the logs were being generated.

Bootstrap DNS is a totally different thing. It's a single DNS query sent by adguard and not logged in filtering logs section. But I get multiple DNS requests leaking to isp DNS visible in filtering logs in adguard, as soon as I open any browser. Did u even look at the logs I sent!! 😐

**Comment by admitrevskiy** [comment]
We seem to have established the cause of the leak. This bug is incredibly difficult to reproduce, because it is not reproducable on most devices. As I've seen in the logs you used Samsung with Android 10, in my case the only one device with the same behavior was Xiaomi Mi 10.

**So, what happens:**
When creating a local VPN we set fake DNS server to reroute all requests to. 
[Documentation](https://developer.android.com/reference/android/net/VpnService.Builder#addDnsServer(java.lang.String)) says: `Adding a server implicitly allows traffic from that address family (i.e., IPv4 or IPv6) to be routed over the VPN.`
However some Android 10 devices routes their traffic via system netd (uid 1051) instead of configured in VPN DNS server.
Henceforward we will start to route all the traffic from port 53 to DnsLibs if DNS module is enabled. Nightly version will be released before long.

**Why did I assume that the problem was caused by a private DNS?**
Private DNS acts very similar: it uses AdGuard DNS as a bootstraper for it's DoT server and there are similar requests in the filtering log. I was a little hasty and made the wrong conclusion, it's my bad.

> It's a single DNS query sent by adguard and not logged in filtering logs section.

Yep, as I've mentioned above `this refers to a single one bootstrap query to resolve DoH/DoT server address` and it was more of a comment about AdGuard's ability to protect you from this type of leak.

I hope our changes will help solve the problem.
